File: kotlin/forth/src/main/kotlin/Forth.kt
class Forth {

    private val stack = mutableListOf<Int>()
    private val definitions = mutableMapOf<String, List<String>>()

    fun evaluate(vararg line: String): List<Int> {

        val tokens = line
            .flatMap { it.lowercase().split("\\s+".toRegex()) }
            .filter { it.isNotBlank() }

        process(tokens)

        return stack.toList()
    }

    private fun process(tokens: List<String>) {
        var i = 0
        while (i < tokens.size) {
            when (val token = tokens[i]) {

                ":" -> {
                    val name = tokens.getOrNull(i + 1)
                        ?: throw IllegalArgumentException("Invalid token $token")
                    require(name.toIntOrNull() == null) { "illegal operation" }

                    val body = mutableListOf<String>()

                    i += 2

                    while (i < tokens.size && tokens[i] != ";") {
                        body += tokens[i]
                        i++
                    }

                    if (i >= tokens.size) {
                        throw IllegalArgumentException("Invalid token $token")
                    }

                    definitions[name] = expand(body)
                }

                else -> execute(token)
            }

            i++
        }
    }

    private fun execute(token: String) {
        token.toIntOrNull()?.let {
            stack.add(it)
            return
        }

        definitions[token]?.let {
            process(it)
            return
        }

        when (token) {
            "+" -> {
                val (a, b) = pop2()
                stack += a + b
            }

            "-" -> {
                val (a, b) = pop2()
                stack += a - b
            }

            "*" -> {
                val (a, b) = pop2()
                stack += a * b
            }

            "/" -> {
                val (a, b) = pop2()
                require(b != 0) { "divide by zero" }
                stack += a / b
            }

            "dup" -> {
                require(stack.isNotEmpty()) { "empty stack" }
                stack += stack.last()
            }

            "drop" -> {
                require(stack.isNotEmpty()) { "empty stack" }
                stack.removeAt(stack.lastIndex)
            }

            "swap" -> {
                require(stack.isNotEmpty()) { "empty stack" }
                require(stack.size >= 2) { "only one value on the stack" }

                val a = stack.removeAt(stack.lastIndex)
                val b = stack.removeAt(stack.lastIndex)

                stack += a
                stack += b
            }

            "over" -> {
                require(stack.isNotEmpty()) { "empty stack" }
                require(stack.size >= 2) { "only one value on the stack" }
                stack += stack[stack.lastIndex - 1]
            }

            else -> throw IllegalArgumentException("undefined operation")
        }
    }

    private fun pop2(): Pair<Int, Int> {
        require(stack.isNotEmpty()) { "empty stack" }
        require(stack.size >= 2) { "only one value on the stack" }

        val b = stack.removeAt(stack.lastIndex)
        val a = stack.removeAt(stack.lastIndex)

        return a to b
    }

    private fun expand(tokens: List<String>): List<String> {
        val result = mutableListOf<String>()

        for (token in tokens) {
            result += definitions[token] ?: listOf(token)
        }

        return result
    }
}

File: csharp/forth/Forth.cs
using System.Text.RegularExpressions;

public static class Forth
{
    private static List<int> stack = [];
    private static Dictionary<string, List<string>> definitions = [];

    public static string Evaluate(string[] instructions)
    {
        stack = [];
        definitions = [];

        var tokens = instructions
            .SelectMany(line => Regex.Split(line.ToLower(), "\\s+"))
            .Where(e => !string.IsNullOrWhiteSpace(e))
            .ToList();

        Process(tokens);

        return string.Join(' ', stack);
    }

    private static void Process(List<string> tokens)
    {
        var i = 0;

        while (i < tokens.Count)
        {
            var token = tokens[i];
            switch (token)
            {
                case ":":
                    var name = i + 1 < tokens.Count ? tokens[i + 1] : throw new ArgumentException($"Invalid token {token}");

                    if (int.TryParse(name, out _))
                    {
                        throw new InvalidOperationException("Illegal operation");
                    }

                    List<string> body = [];

                    i += 2;

                    while (i < tokens.Count && tokens[i] != ";")
                    {
                        body.Add(tokens[i]);
                        i++;
                    }

                    if (i >= tokens.Count) throw new ArgumentException($"Invalid token {token}");

                    definitions[name] = Expand(body);
                    break;
                default:
                    Execute(token);
                    break;
            }

            i++;
        }
    }

    private static void Execute(string token)
    {
        if (int.TryParse(token, out var n))
        {
            stack.Add(n);
            return;
        }

        if (definitions.TryGetValue(token, out var value))
        {
            Process(value);
            return;
        }

        if (token == "+")
        {
            var (a, b) = Pop2();
            stack.Add(a + b);
            return;
        }

        if (token == "-")
        {
            var (a, b) = Pop2();
            stack.Add(a - b);
            return;
        }

        if (token == "*")
        {
            var (a, b) = Pop2();
            stack.Add(a * b);
            return;
        }

        if (token == "/")
        {
            var (a, b) = Pop2();
            if (b == 0) throw new DivideByZeroException();
            stack.Add(a / b);
            return;
        }

        if (token == "dup")
        {
            if (stack.Count == 0) throw new InvalidOperationException("Empty stack");
            stack.Add(stack[^1]);
            return;
        }

        if (token == "drop")
        {
            if (stack.Count == 0) throw new InvalidOperationException("Empty stack");
            stack.RemoveAt(stack.Count - 1);
            return;
        }

        if (token == "swap")
        {
            if (stack.Count == 0) throw new InvalidOperationException("Empty stack");
            if (stack.Count < 2) throw new InvalidOperationException("Only one value on the stack");

            var (a, b) = Pop2();
            stack.Add(b);
            stack.Add(a);

            return;
        }

        if (token == "over")
        {
            if (stack.Count == 0) throw new InvalidOperationException("Empty stack");
            if (stack.Count < 2) throw new InvalidOperationException("Only one value on the stack");

            stack.Add(stack[^2]);

            return;
        }

        throw new InvalidOperationException("Undefined operation");
    }

    private static (int, int) Pop2()
    {
        if (stack.Count <= 0) throw new InvalidOperationException("Empty stack");
        if (stack.Count < 2) throw new InvalidOperationException("Only one value on the stack");

        var b = stack[^1];
        var a = stack[^2];

        stack.RemoveAt(stack.Count - 1);
        stack.RemoveAt(stack.Count - 1);

        return (a, b);
    }

    private static List<string> Expand(List<string> tokens)
    {
        var result = new List<string>();

        foreach (var token in tokens)
        {
            if (definitions.TryGetValue(token, out var value))
            {
                result.AddRange(value);
            }
            else
            {
                result.Add(token);
            }
        }

        return result;
    }
}